Usage alerts

byucesoy opened the pull request:
**Move base_url to config**
Current way of generating base_url depends on the request object. However, we
need base_url in contexts where request object is not available, such as while
sending e-mails from background jobs. With this commit, we make base_url a
config variable. Its default value as http://localhost:9292, which is what we
use for development environment and for production environment we can set it to
https://console.ubicloud.com/.

With this change, we lose the ability to dynamically set the url scheme, but
we redirect http to https in production anyway.

**Move send_email to util**
Previously send_email was part of CloverWeb. This makes it difficult to send an
email in contexts outside of the web application. Sending mail requires setting
up a whole web app, which is not a lightweight operation. The reason of putting
send_email in CloverWeb was to have access to render plugin, because we need to
render html pages as a part of email content. render plugin is part of roda, so
it is not available outside of roda app.

In this commit, I implemented a simple EmailRenderer. It is a wrapper around
Tilt::ErubiTemplate, which is the same as what is used in roda's render plugin.
With the new renderer, we can render templates without having to set up a roda
app and send emails from any context.

This also unblocks sending monthly emails for invoices.

**Add usage alerting based on the total consumption**
This commit adds feature of sending alerts when the total consumption exceeds
a user-defined threshold. We associate each alert to a user who created it and
send the email to that user when the alert is triggered. An alternative would
be allowing users to specify an email address to send the alert to, but that
opens an attack vector where a user could spam another user with alerts. This
vector can be mitigated by requiring the user to confirm the email address, but
implementing a whole confirmation workflow would significantly increase the
effort required to implement this feature.

![Screenshot from 2024-04-14 19-59-58](https://github.com/ubicloud/ubicloud/assets/2082070/2455b2f2-2348-48db-b630-3e579f4148b9)

![Screenshot from 2024-04-15 13-15-28](https://github.com/ubicloud/ubicloud/assets/2082070/d2153974-25b5-4c0d-bffc-a2aa64f54357)

fdr commented on lib/validation.rb:
```diff
@@ -30,6 +30,10 @@ def initialize(details)
 
+  # - Max length 63
+  # - Alphanumeric, hyphen, underscore, space, parantheses, exclamation, question mark, star
+  ALLOWED_SHORT_TEXT_PATTERN = '\A[a-zA-Z0-9_\-!?\*\(\) ]{1,63}\z'
```
dumb question, why are these patterns not Regexp.new / Regexp.compile / %r{} literals that result in a regexp object, rather than strings that presumably have to get compiled over and over?

byucesoy replied:
No idea. Probably the first one is added like this and the rest followed the pattern. I added a separate commit to change them to Regexp.new

fdr commented on ubid.rb:
```diff
@@ -68,6 +68,7 @@ class UBID
   TYPE_FIREWALL_RULE = "fr"
   TYPE_FIREWALL = "fw"
   TYPE_POSTGRES_FIREWALL_RULE = "pf"
+  TYPE_USAGE_ALERT = "a1"
```
Seems unfortunate to sit on what is presumably an abbreviation of "alert" (a1ert) for _usage_ records only. Seems hardly the most important or dominant kind of alert, if we want to build an alerting channel, which we probably will at some point. Though I suppose you could try to pivot the meaning (e.g. make UsageAlert a join on an Alert table and backfill) to rectify it post-facto.

byucesoy replied:
I changed it to va as abbreviation of "vsage alert"

fdr commented:
One thing that is kinda bothering me about this patch is that it opens a significant new territory to administer and adjust. And that is:

1. Alerting via email as a general matter
2. A new feature given top level UI footprint along with VMs, Networking, Billing, Postgres, etc
3. The feature is customarily seen as part of a budget management feature

To me, it seems like opening a problem prematurely, that is, people are going to click that tab and ask for more adjustments. Is there a theory of evolution here? Who is administering the feedback for it and saying "nah we're not going to do that" or "yeah okay it fits into our general theory like this"? Do we want to feature flag this thing? Does it need APIs as well or is it a UI only feature (which is a bit smelly).

fdr commented:
Some other alternatives:

1. Make it a box or two in the billing page. This invites fewer adjustment requests.
2. Attack the email alert channel generally, streamlining inefficient aspects of our process that currently involves manual email.

Examples of the second: sending emails about credit card declined, suspension for ToS, or pending maintenance/downtime.  Here we could book a win if the cost or administering the alerting system and whatever appendages it grows into the UI/API is lower than the frictions caused by the status quo processes.

It is a bit of mission creep, which can be elided if you are confident in how to keep the system compact while adding new alert channels. But if you are not confident, some level of prototyping may be indicated.

byucesoy commented:
> Some other alternatives:
> 1. Make it a box or two in the billing page. This invites fewer adjustment requests, I think, than conveying this as a Feature at the same level as "VM", "Settings" or "Postgres"
> 2. Attack the email alert channel generally, streamlining inefficient aspects of our process that currently involves manual email.

Both makes sense. I moved removed the top level entry and moved usage alerts to the billing page. In this way, I think it wouldn't be seen as first class feature until we want to make it so.

On second one, when I think about alerts and emails, I see lots of expansion points:
- Disk, memory, cpu usage alerts
- Getting alerts via other means such as SMS or pinging 3rd party endpoints
- Maintenances, invoices, account suspensions (i.e. all the stuff we do manually right now)

However, I don't think we need to attack those right now. Being able to get alerts for consumption seems like a good enough start and also gives us some opportunity to get our feet wet.

> Note that you become in charge of the software architecture evolution of the email alert system generally by default if you do this :)

I'm fine with that :smile:.

fdr commented on ubid.rb:
```diff
@@ -68,6 +68,7 @@ class UBID
   TYPE_FIREWALL_RULE = "fr"
   TYPE_FIREWALL = "fw"
   TYPE_POSTGRES_FIREWALL_RULE = "pf"
+  TYPE_USAGE_ALERT = "va"
```
Should we be making ubid types for every kind of alert or should there be TYPE_ALERT and it's polymorphic in some other way?

byucesoy replied:
I decided to use UBID::TYPE_ETC

enescakir commented on config.rb:
```diff
@@ -63,6 +63,7 @@ def self.test?
   override :smtp_tls, true, bool
 
   # Override -- value is returned or the set default.
+  override :base_url, "http://localhost:9292", string
```
We have hardcoded `https://console.ubicloud.com` strings in the codebase. Could you replace them as well?

enescakir commented on lib/util.rb:
```diff
@@ -74,4 +77,28 @@ def self.safe_write_to_file(filename, content)
+
+  def self.send_email(receiver, subject, greeting: nil, body: nil, button_title: nil, button_link: nil)
+    html = EmailRenderer.new.render "email/layout", locals: {subject: subject, greeting: greeting, body: body, button_title: button_title, button_link: button_link}
+    Mail.deliver do
```
You can set it in the `Mail.deliver` or you can set it at more common place than Roda.
```
      case Config.mail_driver
      when :smtp
        delivery_method :smtp, {
          address: Config.smtp_hostname,
          port: Config.smtp_port,
          user_name: Config.smtp_user,
          password: Config.smtp_password,
          authentication: :plain,
          enable_starttls: Config.smtp_tls
        }
      when :logger
        delivery_method :logger
      when :test
        delivery_method :test
      end
```

enescakir commented on model/usage_alert.rb:
```diff
@@ -0,0 +1,25 @@
+      button_link: "#{Config.base_url}#{project.path}/dashboard")
+  end
+
+  include ResourceMethods
```
Should we include modules before any methods, like we do with other models?

enescakir commented on model/usage_alert.rb:
```diff
@@ -0,0 +1,25 @@
+      body: ["The usage alert, #{name}, you set for project #{project.name} (id: #{project.ubid}) has been triggered.",
+        "Current cost: #{project.current_invoice.content["cost"].to_f.round(2)}",
+        "Please note that this alert is only for informational purposes and no action is taken automatically."],
+      button_title: "Go to Project",
+      button_link: "#{Config.base_url}#{project.path}/dashboard")
```
Should we redirect to the billing page instead of the dashboard?
```suggestion
      button_title: "See usage",
      button_link: "#{Config.base_url}#{project.path}/billing")
```

enescakir commented on model/usage_alert.rb:
```diff
@@ -0,0 +1,25 @@
+    Util.send_email(user.email, "Usage alert is triggered for project #{project.name}",
+      greeting: "Hello #{user.name},",
+      body: ["The usage alert, #{name}, you set for project #{project.name} (id: #{project.ubid}) has been triggered.",
+        "Current cost: #{project.current_invoice.content["cost"].to_f.round(2)}",
```
```suggestion
        "Current cost: $#{project.current_invoice.content["cost"].to_f.round(2)}",
```

enescakir commented on routes/web/project/billing.rb:
```diff
@@ -19,6 +19,8 @@ class CloverWeb
         @invoices = Serializers::Web::Invoice.serialize(@project.invoices.prepend(@project.current_invoice))
       end
 
+      @usage_alerts = Serializers::Web::UsageAlert.serialize(@project.usage_alerts)
```
We can load users eagerly.

enescakir commented on views/project/billing.erb:
```diff
@@ -301,3 +301,108 @@
+              </p>
+            </div>
+            <div class="grid grid-cols-12 gap-6">
+              <div class="col-span-5 sm:col-span-3">
+                <%== render(
+                  "components/form/text",
+                  locals: {
+                    name: "alert_name",
+                    label: "Alert Name",
+                    type: "text",
+                    attributes: {
+                      required: true,
+                      placeholder: "Alert Name"
+                    },
+                    extra_class: "pr-3"
+                  }
+                ) %>
+              </div>
+              <div class="col-span-5 sm:col-span-3">
+                <%== render(
+                  "components/form/text",
+                  locals: {
+                    name: "limit",
+                    type: "number",
+                    attributes: {
+                      required: true,
+                      placeholder: "100"
+                    },
+                    extra_class: "pr-3"
+                  }
+                ) %>
+              </div>
```
Can you add a label to limit too or remove the label from name? They look unaligned vertically.

enescakir commented on views/project/billing.erb:
```diff
@@ -301,3 +301,108 @@
+          <thead class="bg-gray-50">
+            <tr>
+              <th scope="col" class="py-3.5 pl-4 pr-3 text-left text-sm font-semibold text-gray-900 sm:pl-4">Name</th>
+              <th scope="col" class="py-3.5 pl-4 pr-3 text-left text-sm font-semibold text-gray-900 sm:pl-4">Limit</th>
```
Since, multiple users can access the billing page and add alerts. Should we add also email column here?

fdr reviewed (approved):
Modulo @enescakir's find, looks like we have a second, probably more inelastic requirement for email sending: invoices, to help with VAT enforcement.

I'm a bit concerned about proliferating a lot of ubid types, one for each alert type: it seems like it'll mushroom into too many types. Addressing how to do ubid types is at least a little important, because depending how the program evolves, it can be expensive to change. Measure twice, here.

I think my requests for considerations have been absorbed and I'll leave the rest to programmer prerogative for adjustments.

enescakir commented on views/components/icon.erb:
```diff
@@ -113,6 +113,8 @@
   <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="<%= classes %>">
     <path stroke-linecap="round" stroke-linejoin="round" d="M16.5 10.5V6.75a4.5 4.5 0 10-9 0v3.75m-.75 11.25h10.5a2.25 2.25 0 002.25-2.25v-6.75a2.25 2.25 0 00-2.25-2.25H6.75a2.25 2.25 0 00-2.25 2.25v6.75a2.25 2.25 0 002.25 2.25z" />
   </svg>
+<% when "hero-bell-alert" %>
```
also .svg file

byucesoy commented:
For the ubid, I used UBID::TYPE_ETC to not overcrowd the global ubid namespace.
